lib/YuiRestClient/App.pm

```diff
@@ -55,6 +55,7 @@ sub get_host {
 
 sub check_connection {
```

Comment:
I agree that the function it is not well done, so after checking the boolean, we would need another method to retrieve the page and get the current json. That makes sense.
Regarding the wrapping of wait_until in (2) if we do always, we need to adjust each call to die when some page is not found. For that reason we could create some method like `try(or check)_wait_until` which would return boolean.
Instead of hacking those `die`s allowing both 'reporting behaviours' it could be done with eval, but it would be an eval inside another eval (I'm not sure how good is that and how obscure is debug it).

Reply:
Hey @ge0r @OleksandrOrlov and me discussed offline and we think:
- (1) we need to improve that method, return the json if succeded and undef if not, and handle properly in setup_libyui and in this particular case for post_fail_hook.
- (2) we create a `try_wait_until` which wraps `wait_until` with eval and returns boolean value.
And the use this last function inside check_connection.

Reply:
> Hey @ge0r @OleksandrOrlov and me discussed offline and we think:
> 
>     * (1) we need to improve that method, return the json if succeded and undef if not, and handle properly in setup_libyui and in this particular case for post_fail_hook.
so if `check_connection` fails, it returns undef, and then `Module.pm` `y2base.pm` and `setup_libyui.pm` will either die (because right now some these modules expect to die if there is no connection to the libyui server) or handle it, depending on the case.

>     * (2) we create a `try_wait_until` which wraps `wait_until` with eval and returns boolean value.
>       And the use this last function inside check_connection.
the thing is, in case eg `setup_libyui.pm` has `check_connection` failing, currently it dies. Isn't it verbose to call the `try_wait_until` via `check_connection` and to handle the error on the setup_libyui level? Ideally for this scenario we should use the `wait_until` which dies with the appropriate error, not `try_wait_until`

Reply:
Otherwise, for consistency, all functions should call `try_wait_until` and do whatever they want based on the result (including die), and `wait_until` should be deleted.

Reply:
It would be more consistent to have those functions as they are, and catch the die exception, whenever we have to handle it, on the level of the modules that call check_connection.

Reply:
seems that this is needed in multiple places, just noticed that we are already using it:

```perl
sub is_shown {
    my ($self) = @_;
    my $result;
    eval {
        $result = YuiRestClient::Wait::wait_until(object => sub {
                return $self->{lbl_wizard}->property('debug_label') eq 'Select Hard Disk(s)';
        });
    };
    $result ? 1 : 0;
}
```
This could be simplified to:
```perl

sub is_shown {
    my ($self) = @_;
    YuiRestClient::Wait::try_wait_until(object => sub {
                $self->{lbl_wizard}->property('debug_label') eq 'Select Hard Disk(s)';
        });
}
```
I have doubts about silent post_fail_hooks, I'm not sure how is the more realiable way, that is something we should explore in those ticket to be created that we were talking offline. Otherwise on timeouts that way of coding could be propagated and really hard to debug.

`check_connection` for the name of the method I would say it shouldn't die, when if I would read `connect` I would assume it could die, but check can be understood like true or false. 

In other frameworks I was used to have two versions of the functions sometimes, they were prepended with Try.
If we create a function that does it on purpose is more clear. We need the default behaviour to not being creating dies everywhere with the same message of timeout, but there are a few exception that we need to deal with like this. From the user perspective there is to functions documented, I don't see any problem.